core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ParentComposition.kt

```diff
@@ -50,31 +48,29 @@ internal data class ParentComposition(
  * Creates a [ParentComposition] from the current point in the composition and adds it to this
  * [ViewEnvironment].
  */
-@Composable internal fun ViewEnvironment.withParentComposition(): ViewEnvironment {
-  val compositionReference = ParentComposition(
-      reference = compositionReference(),
-      recomposer = currentComposer.recomposer
-  )
+@Composable internal fun ViewEnvironment.withParentComposition(
+  reference: CompositionReference = compositionReference()
+): ViewEnvironment {
+  val compositionReference = ParentComposition(reference = reference)
   return this + (ParentComposition to compositionReference)
 }
 
 /**
  * Starts composing [content] into this [ViewGroup].
  *
- * If there is a [ParentComposition] present in [initialViewEnvironment], it will start the
- * composition as a subcomposition of that continuation.
+ * If [parentComposition] is not null, [content] will be installed as a _subcomposition_ of the
+ * parent composition, meaning that it will propagate ambients and invalidation.
  *
  * This function corresponds to [withParentComposition].
  */
-internal fun ViewGroup.setOrContinueContent(
-  initialViewEnvironment: ViewEnvironment,
+internal fun ViewGroup.setOrSubcomposeContent(
+  parentComposition: CompositionReference?,
   content: @Composable() () -> Unit
 ) {
-  val (compositionReference, recomposer) = initialViewEnvironment[ParentComposition]
-  if (compositionReference != null && recomposer != null) {
+  if (parentComposition != null) {
     // Somewhere above us in the workflow rendering tree, there's another bindCompose factory.
     // We need to link to its composition reference so we inherit its ambients.
-    setContent(recomposer, compositionReference, content)
+    setContent(Recomposer.current(), parentComposition, content)
```

Comment:
Everything is just using the main thread Recomposer, no need to pass it around ourselves.